### PR TITLE
[CDAP-20990 ] Added Repo Source control metadata store methods

### PR DESCRIPTION
**Pull request by adrikagupta**
- Added methods (write and delete) required for repo source control metadata
- Added unit tests for the same

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/SourceControlMetadataStore.java [hunk 1]
```diff
@@ -64,6 +69,18 @@ private StructuredTable getNamespaceSourceControlMetadataTable() {
     return namespaceSourceControlMetadataTable;
   }
 
+  private StructuredTable getRepositorySourceControlMetadataTable() {
+    try {
+      if (repositorySourceControlMetadataTable == null) {
+        repositorySourceControlMetadataTable = context.getTable(
+            StoreDefinition.RepositorySourceControlMetadataStore.REPOSITORY_SOURCE_CONTROL_METADATA);
+      }
+    } catch (TableNotFoundException e) {
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/SourceControlMetadataStore.java]: throw `TableNotFoundException` directly as it is already `RuntimeException`

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/SourceControlMetadataStore.java [hunk 2]
```diff
@@ -121,6 +138,22 @@ public void write(ApplicationId appId,
...
+   * Writes or updates the source control metadata for the specified repository metadata ID.
+   *
+   * @param repoMetadataId The {@link RepoMetadataId} for which source control metadata needs to be
+   *                       written or updated.
+   * @param isSynced       A boolean indicating whether the entity in repository is synced with the
+   *                       one in namespace
+   * @param lastSynced     The timestamp of the last git pull/push of the entity
+   * @throws IOException If an I/O error occurs while writing the source control metadata.
+   */
+  public void writeRepoSourceControlMeta(RepoMetadataId repoMetadataId, Boolean isSynced,
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/SourceControlMetadataStore.java]: Use `ApplicationReference` than `RepoMetadataId` as `ApplicationReference` identifies versionless application.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/SourceControlMetadataStore.java [hunk 3]
```diff
@@ -142,6 +175,18 @@ public void deleteAll(String namespace) throws IOException {
...
+   * @param repoMetadataId The {@link RepoMetadataId} for which source control metadata needs
+   *                       to be deleted.
+   * @throws IOException If an I/O error occurs while deleting the source control metadata.
+   */
+  public void deleteRepoSourceControlMeta(RepoMetadataId repoMetadataId) throws IOException {
+    getRepositorySourceControlMetadataTable().delete(
+        getRepoSourceControlMetadataPrimaryKey(repoMetadataId));
+  }
+
   private Collection<Field<?>> getNamespaceSourceControlMetaFields(ApplicationId appId,
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/SourceControlMetadataStore.java]: Use `ApplicationReference` than `ApplicationId` as `ApplicationReference` identifies versionless application.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/SourceControlMetadataStore.java (hunk 1 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/SourceControlMetadataStore.java]: Fixed. Thanks

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/SourceControlMetadataStore.java (hunk 3 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/SourceControlMetadataStore.java]: Right. Thank you for this correction. Fixed

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/SourceControlMetadataStore.java (hunk 2 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/SourceControlMetadataStore.java]: Fixed

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import javax.annotation.Nullable;
+
+/**
+ * Store for namespace source control metadata.
+ */
+public class NamespaceSourceControlMetadataStore {
```
Comment (adrikagupta) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java:43]: Renamed SourceControlMetadataStore to NamespaceSourceControlMetadataStore

Comment (adrikagupta) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java:42]: Renamed SourceControlMetadataStore to NamespaceSourceControlMetadataStore

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+            scmMeta == null ? "" : scmMeta.getCommitId()));
+    // Whenever an app is deployed, the expected behavior is that the last modified field will be
+    // retained and not reset.
+    Long lastModified = 0L;
+    if (scmMeta != null) {
+      lastModified = scmMeta.getLastSyncedAt().toEpochMilli();
+    } else {
+      SourceControlMeta sourceControlMeta = get(appRef);
+      if (sourceControlMeta != null) {
+        lastModified = sourceControlMeta.getLastSyncedAt().toEpochMilli();
```
Comment (adrikagupta) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java]: @samdgupi 
We are getting lastSynced as Instant in SourceControlMeta and I have converted it using toEpochMilli()
Let me know if its okay or do we need to use getEpochSecond or some other function

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+    List<Field<?>> fields = getPrimaryKey(appRef);
+    fields.add(Fields.stringField(
+        StoreDefinition.NamespaceSourceControlMetadataStore.SPECIFICATION_HASH_FIELD,
+        scmMeta == null ? "" : scmMeta.getFileHash()));
+    fields.add(
+        Fields.stringField(StoreDefinition.NamespaceSourceControlMetadataStore.COMMIT_ID_FIELD,
+            scmMeta == null ? "" : scmMeta.getCommitId()));
+    // Whenever an app is deployed, the expected behavior is that the last modified field will be
+    // retained and not reset.
+    Long lastModified = 0L;
```
Comment (adrikagupta) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java]: @samdgupi 
By default when an app is deployed and not already present in the new table, the value for last modified is kept as 0 and not null. I did this cause rows can be sorted on this field. Just wanted to let you know explicitly. Let me know if any changes are needed

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/DefaultStore.java [hunk 7]
```diff
@@ -601,15 +602,18 @@ public void markApplicationsLatest(Collection<ApplicationId> appIds)
...
     return TransactionRunners.run(transactionRunner, context -> {
-      getSourceControlMetadataStore(context).write(id, meta.getSourceControlMeta());
+      getNamespaceSourceControlMetadataStore(context).write(id.getAppReference(),
+          meta.getSourceControlMeta());
       return getAppMetadataStore(context).createLatestApplicationVersion(id, meta);
     }, ConflictException.class);
   }
 
   @Override
-  public int addApplication(ApplicationId id, ApplicationMeta meta, boolean isLatest) throws ConflictException {
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/DefaultStore.java:610]: Please revert this formatting change

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java (hunk 5 above)
Comment (samdgupi) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java]: toEpochMilli should be ok

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/DefaultStore.java (hunk 7 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/DefaultStore.java:610]: Done

File: cdap-proto/src/main/java/io/cdap/cdap/proto/sourcecontrol/SourceControlMeta.java [hunk 8]
```diff
@@ -29,18 +29,41 @@ public class SourceControlMeta {
...
+   * @param fileHash     The hash value of the file associated with the metadata.
+   * @param commitId     The ID of the commit associated with the metadata, or {@code null} if not
+   *                     available.
+   * @param lastSyncedAt The timestamp when the metadata was last synced, or {@code null} if not
+   *                     available.
+   * @param syncStatus   The synchronization status indicating whether the metadata is synchronized,
+   *                     represented as a boolean value.
+   */
+  public SourceControlMeta(String fileHash, @Nullable String commitId,
+      @Nullable Instant lastSyncedAt, Boolean syncStatus) {
```
Comment (samdgupi) [inline cdap-proto/src/main/java/io/cdap/cdap/proto/sourcecontrol/SourceControlMeta.java:62]: syncstatus should be nullable

Comment (samdgupi) [inline cdap-proto/src/main/java/io/cdap/cdap/proto/sourcecontrol/SourceControlMeta.java]: sync status should be nullable

File: cdap-proto/src/main/java/io/cdap/cdap/proto/sourcecontrol/SourceControlMeta.java [hunk 9]
```diff
@@ -29,18 +29,41 @@ public class SourceControlMeta {
   private final String commitId;
   // The last time the application was synced(push/pull) with git.
   private final Instant lastSyncedAt;
+  @Nullable
```
Comment (samdgupi) [inline cdap-proto/src/main/java/io/cdap/cdap/proto/sourcecontrol/SourceControlMeta.java:32]: where is getter ?
Also this needs to be added in hash and equals

File: cdap-proto/src/main/java/io/cdap/cdap/proto/sourcecontrol/SourceControlMeta.java (hunk 8 above)
Comment (adrikagupta) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/sourcecontrol/SourceControlMeta.java]: done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+    StructuredTable scmTable = getNamespaceSourceControlMetadataTable();
+    SourceControlMeta existingSourceControlMeta = get(appRef);
+    if (sourceControlMeta != null && existingSourceControlMeta != null
+        && sourceControlMeta.getLastSyncedAt()
+        .isBefore(existingSourceControlMeta.getLastSyncedAt())) {
+      List<Field<?>> fields = new ArrayList<>();
+      fields.add(
+          Fields.longField(StoreDefinition.NamespaceSourceControlMetadataStore.LAST_MODIFIED_FIELD,
+              sourceControlMeta.getLastSyncedAt().toEpochMilli()));
+      throw new InvalidFieldException(
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java]: I dont think `InvalidFieldException` is meant for this usecase. Use `IllegalArgumentException` instead. Also do not need to create a field object for this case. Just write a message with the details

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+   *
+   * @param appRef            {@link ApplicationReference} for which the source control metadata is
+   *                          being set.
+   * @param sourceControlMeta The {@link SourceControlMeta} to be set. Can be {@code null} if
+   *                          application is just deployed.
+   * @throws IOException If failed to write the data.
+   */
+  public void write(ApplicationReference appRef,
+      @Nullable SourceControlMeta sourceControlMeta)
+      throws IOException {
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java]: Add `IllegalArgumentException` to the throws list

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+        StoreDefinition.NamespaceSourceControlMetadataStore.SPECIFICATION_HASH_FIELD,
+        newSourceControlMeta == null ? null : newSourceControlMeta.getFileHash()));
+    fields.add(
+        Fields.stringField(StoreDefinition.NamespaceSourceControlMetadataStore.COMMIT_ID_FIELD,
+            newSourceControlMeta == null ? null : newSourceControlMeta.getCommitId()));
+    // Whenever an app is deployed, the expected behavior is that the last modified field will be
+    // retained and not reset.
+    fields.add(
+        Fields.longField(StoreDefinition.NamespaceSourceControlMetadataStore.LAST_MODIFIED_FIELD,
+            newSourceControlMeta == null ? (existingSourceControlMeta == null ? 0L
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java]: null rather than `oL`

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+                : existingSourceControlMeta.getLastSyncedAt().toEpochMilli())
+                : newSourceControlMeta.getLastSyncedAt().toEpochMilli()));
+    fields.add(
+        Fields.booleanField(StoreDefinition.NamespaceSourceControlMetadataStore.IS_SYNCED_FIELD,
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java:177]: hmm now that I think about it you can always send a sourcecontrol meta. just the sycned field will be false. That is the base state when a app is added. That way you do not have to deal with null source control meta at all.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleServiceTest.java [hunk 14]
```diff
@@ -757,6 +752,9 @@ public void testUpdateSourceControlMetaWithNonExistingApp() throws Exception {
     Assert.assertNotNull(updatedDetail.getSourceControlMeta());
     Assert.assertEquals("updated-file-hash", updatedDetail.getSourceControlMeta().getFileHash());
     Assert.assertEquals("updated-commit-id", updatedDetail.getSourceControlMeta().getCommitId());
+
+    deleteAppAndData(new ApplicationId(TEST_NAMESPACE1, AllProgramsApp.NAME,
```
Comment (samdgupi) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleServiceTest.java:756]: why this change?

Comment (adrikagupta) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleServiceTest.java:756]: While running ApplicationLifecycleServiceTest, in this specific test we are deploy an app and updating its source control metadata. In another test, testUpdateSourceControlMeta, we deploy the same app. Its filehash and commitId becomes null but the last modified retains (it is the expected behaviour now), hence source control metadata is not null. 
I made this change so that a unit test doesn't affect another test.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java (hunk 13 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java:177]: Right. Done. Thank you for pointing that out

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java (hunk 11 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java]: Done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java (hunk 10 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java]: Done

File: cdap-proto/src/main/java/io/cdap/cdap/proto/sourcecontrol/SourceControlMeta.java (hunk 9 above)
Comment (adrikagupta) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/sourcecontrol/SourceControlMeta.java:32]: Done

File: cdap-proto/src/main/java/io/cdap/cdap/proto/sourcecontrol/SourceControlMeta.java [hunk 15]
```diff
@@ -29,18 +29,41 @@ public class SourceControlMeta {
...
-   * @param fileHash the git-hash of the config in git after push
-   * @param commitId the commit in git form/to application was pulled/pushed
+   * @param fileHash     the git-hash of the config in git after push
+   * @param commitId     the commit in git form/to application was pulled/pushed
    * @param lastSyncedAt last time the application was pulled/pushed
    */
-  public SourceControlMeta(String fileHash, @Nullable String commitId, @Nullable Instant lastSyncedAt) {
+  public SourceControlMeta(String fileHash, @Nullable String commitId,
+      @Nullable Instant lastSyncedAt) {
+    this.fileHash = fileHash;
```
Comment (samdgupi) [inline cdap-proto/src/main/java/io/cdap/cdap/proto/sourcecontrol/SourceControlMeta.java]: call `SourceControlMeta(String fileHash, @Nullable String commitId,
      @Nullable Instant lastSyncedAt, @Nullable Boolean syncStatus)` with null instead

File: cdap-proto/src/main/java/io/cdap/cdap/proto/sourcecontrol/SourceControlMeta.java [hunk 16]
```diff
@@ -86,4 +108,22 @@ public String getCommitId() {
...
+  @Nullable
+  public Boolean getSyncStatus() {
+    return syncStatus;
+  }
+
+  public static SourceControlMeta createDefaultMeta() {
+    return new SourceControlMeta(null, null, null, false);
+  }
+
+  public SourceControlMeta withUpdatedSyncStatus(Boolean newSyncStatus) {
```
Comment (samdgupi) [inline cdap-proto/src/main/java/io/cdap/cdap/proto/sourcecontrol/SourceControlMeta.java]: Use builder pattern. Only allow setSyncStatus method

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java (hunk 11 above)
Comment (samdgupi) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java]: I dont see the change

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java [hunk 17]
```diff
@@ -142,46 +158,69 @@ public void deleteAll(String namespace) throws IOException {
     getNamespaceSourceControlMetadataTable().deleteAll(getNamespaceRange(namespace));
   }
 
-  private Collection<Field<?>> getNamespaceSourceControlMetaFields(ApplicationId appId,
-      SourceControlMeta scmMeta) throws IOException {
-    List<Field<?>> fields = getPrimaryKey(appId);
+//  private Collection<Field<?>> getAllFields(ApplicationReference appRef,
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java]: why is this commented?

File: cdap-proto/src/main/java/io/cdap/cdap/proto/sourcecontrol/SourceControlMeta.java (hunk 15 above)
Comment (adrikagupta) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/sourcecontrol/SourceControlMeta.java]: From what I am interepreting, instead of doing this.syncstatus = null, I should pass it in the constructor.
This would require change in existing code like pullAndDeploy, deploy, pushApp, updateSourceControlMetadata functions and corresponding tests. Is that fine?

File: cdap-proto/src/main/java/io/cdap/cdap/proto/sourcecontrol/SourceControlMeta.java (hunk 16 above)
Comment (adrikagupta) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/sourcecontrol/SourceControlMeta.java]: Are you suggesting this?
` meta.getSourceControlMeta() == null ? SourceControlMeta.createDefaultMeta()
              : SourceControlMeta.builder(meta.getSourceControlMeta().getFileHash(),
                  meta.getSourceControlMeta().getCommitId(),
                  meta.getSourceControlMeta().getLastSyncedAt()).setSyncStatus(true).build()`

File: cdap-proto/src/main/java/io/cdap/cdap/proto/sourcecontrol/SourceControlMeta.java (hunk 15 above)
Comment (samdgupi) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/sourcecontrol/SourceControlMeta.java]: What I meant you can call `SourceControlMeta(String fileHash, @Nullable String commitId, @Nullable Instant lastSyncedAt, @Nullable Boolean syncStatus)` from inside `SourceControlMeta(String fileHash, @Nullable String commitId, @Nullable Instant lastSyncedAt)` rather than setting fields individually.

File: cdap-proto/src/main/java/io/cdap/cdap/proto/sourcecontrol/SourceControlMeta.java (hunk 16 above)
Comment (samdgupi) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/sourcecontrol/SourceControlMeta.java]: See `io/cdap/cdap/proto/operation/OperationMeta.java:89` on how to use a builder pattern. For builder constructor you can only support building from another SourceControlMeta.

```
 public static Builder builder() {
    return new Builder();
  }

  public static Builder builder(SourceControlMeta meta) {
    return new Builder()
        // call set methods
  }

  @SuppressWarnings("unchecked")
  public static class Builder {

    // fields

    protected Builder() {
      // initial state if needed
    }
     
     // set methods

    public OperationMeta build() {
      // any checks if needed and then return SourceControlMeta
  }
}

```

File: cdap-proto/src/main/java/io/cdap/cdap/proto/sourcecontrol/SourceControlMeta.java (hunk 15 above)
Comment (adrikagupta) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/sourcecontrol/SourceControlMeta.java]: Thanks for the clarification. Done

File: cdap-proto/src/main/java/io/cdap/cdap/proto/sourcecontrol/SourceControlMeta.java (hunk 16 above)
Comment (adrikagupta) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/sourcecontrol/SourceControlMeta.java]: Thanks for the clarification. Done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java (hunk 17 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java]: My bad. Removed it

**Review by samdgupi — APPROVED** [review]
